appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockContainer.java

```diff
@@ -318,8 +318,16 @@ public final void onDrop(DragSource source, int x, int y, int offsetX, int offse
   @Override
   public void onRemoved()
   {
-    for (MockComponent child : children) {
+    // Traverse list backwards to make removal easier
+    for (int i = children.size() - 1; i >= 0; --i) {
+      MockComponent child = children.get(i);
+
+      // Send event of component removal
       getForm().fireComponentRemoved(child, true);
```

Comment:
It seems like this would result in `fireComponentRemoved` to get called twice. Once here and once in `removeComponent` which will be called by `child.delete()`. I haven't yet tried logging this to confirm.

Reply:
> It seems like this would result in `fireComponentRemoved` to get called twice. Once here and once in `removeComponent` which will be called by `child.delete()`. I haven't yet tried logging this to confirm.

Confirmed.Maybe we should overwrite delete() to delete all children before the container is deleted.

Reply:
```
  @Override
  public void delete() {
    // Traverse list backwards to make removal easier
    for (int i = children.size() - 1; i >= 0; --i) {
      // Manually delete child component to ensure that it is
      // completely removed from the Designer.
      children.get(i).delete();
    }
    super.delete();
  }
```

Reply:
Good catch!

@wxbit I believe the removal of the getForm.fireComponentRemoved() line should be sufficient. I think overriding delete vs overriding onRemoved should give the same result, but I chose to override onRemoved since it was already implemented to also remove all the children (by sending the event prior to the implementation)

Reply:
No, it's different. onRemoved is invoked after the container is deleted, MUST overwrite delete..

Reply:
@wxbit You are right, I realized it might make a difference if the child component is somehow using the parent component in between, so it might make more sense to delete the children before deleting the container. I have made the change, cheers for the suggestion!